include/oneapi/dpl/internal/distributed_ranges_impl/sp/algorithms/reduce.hpp

```diff
@@ -0,0 +1,190 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _ONEDPL_DR_SP_REDUCE_HPP
+#define _ONEDPL_DR_SP_REDUCE_HPP
+
+#include <sycl/sycl.hpp>
+
+#include "oneapi/dpl/async"
+#include "oneapi/dpl/execution"
+#include "oneapi/dpl/numeric"
+
+#include "../../concepts/concepts.hpp"
+#include "../../detail/onedpl_direct_iterator.hpp"
+#include "../init.hpp"
+#include "execution_policy.hpp"
+
+namespace
+{
+
+// Precondition: stdrng::distance(first, last) >= 2
+// Postcondition: return future to [first, last) reduced with fn
+template <typename T, typename ExecutionPolicy, std::bidirectional_iterator Iter, typename Fn>
+auto
+reduce_no_init_async(ExecutionPolicy&& policy, Iter first, Iter last, Fn&& fn)
```

Comment:
Should we uglify these internal helpers since they shouldn't be called by a user?

Reply:
I think it is enough to hide the names in an internal namespace, without uglification. Overall. uglification is needed to avoid name clashes, not to indicate internal status of certain functions.